baseapp/protocompat/protocompat.go

```diff
@@ -0,0 +1,191 @@
+package protocompat
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	gogoproto "github.com/cosmos/gogoproto/proto"
+	"google.golang.org/grpc"
+	proto2 "google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/reflect/protoregistry"
+)
+
+var (
+	gogoType    = reflect.TypeOf((*gogoproto.Message)(nil)).Elem()
+	protov2Type = reflect.TypeOf((*proto2.Message)(nil)).Elem()
+)
+
+type Handler = func(ctx context.Context, message gogoproto.Message) (gogoproto.Message, error)
+
+func MakeHybridHandler(cdc codec.BinaryCodec, sd *grpc.ServiceDesc, method grpc.MethodDesc, handler interface{}) (Handler, error) {
+	methodFullName := protoreflect.FullName(fmt.Sprintf("%s.%s", sd.ServiceName, method.MethodName))
+	desc, err := gogoproto.HybridResolver.FindDescriptorByName(methodFullName)
+	if err != nil {
+		return nil, err
+	}
+	methodDesc, ok := desc.(protoreflect.MethodDescriptor)
+	if !ok {
+		return nil, fmt.Errorf("invalid method descriptor %s", methodFullName)
+	}
+
+	funcType, err := getFuncType(handler, method.MethodName)
+	if err != nil {
+		return nil, err
+	}
+	// the handler function takes two arguments: context.Context and proto.Message
+	// since it is a method, the first argument is the receiver, which we don't care about.
+	// the second argument is the proto.Message, which we need to know the type of.
+	inputType := funcType.In(2)
+	switch {
+	case inputType.Implements(protov2Type):
+		return makeProtoV2HybridHandler(methodDesc, cdc, method, handler)
+	case inputType.Implements(gogoType):
+		return makeGogoHybridHandler(methodDesc, cdc, method, handler)
+	default:
+		return nil, fmt.Errorf("invalid method handler type %T, input does not implement", method.Handler)
+	}
+}
+
+// makeProtoV2HybridHandler returns a handler that can handle both gogo and protov2 messages.
+func makeProtoV2HybridHandler(prefMethod protoreflect.MethodDescriptor, cdc codec.BinaryCodec, method grpc.MethodDesc, handler any) (Handler, error) {
+	// it's a protov2 handler, if a gogo counterparty is not found we cannot handle gogo messages.
+	gogoRespType := gogoproto.MessageType(string(prefMethod.Output().FullName()))
+	if gogoRespType == nil {
+		return func(ctx context.Context, request gogoproto.Message) (gogoproto.Message, error) {
+			protov2Request, ok := request.(proto2.Message)
+			if !ok {
+				return nil, fmt.Errorf("invalid request type %T, method %s does not accept gogoproto messages", request, prefMethod.FullName())
+			}
+			resp, err := method.Handler(handler, ctx, func(msg any) error {
+				proto2.Merge(msg.(proto2.Message), protov2Request)
+				return nil
+			}, nil)
+			if err != nil {
+				return nil, err
+			}
+			return resp.(gogoproto.Message), nil
+		}, nil
+	}
+	gogoRespType = gogoRespType.Elem() // we need the non pointer type
+	return func(ctx context.Context, request gogoproto.Message) (gogoproto.Message, error) {
```

Comment:
we discussed in today's meeting, could this interface be used https://pkg.go.dev/google.golang.org/protobuf/runtime/protoiface#MessageV1 instead of `gogoproto.Message`?